### PR TITLE
Make links in todo messages consistent

### PR DESCRIPTION
**Pull request by checkaayush**
#### Summary
Makes links in todo messages consistent.

- For unread messages:
  - `[org/repo-last-token-ellipsis](repoURL) <type> [title](issueURL)`
  - `<type>` above will be either `PullRequest` or `Issue`
  - Example: 
    - [mattermost/...server](https://github.com/mattermost/mattermost-server) Issue [Rewrite existing plugin API tests to exclusively test via actual plugin](https://github.com/mattermost/mattermost-server/issues/10438)

- For other categories:
  - `[org/repo-last-token-ellipsis](repoURL) [title](issueURL)`
  - Example:
    - [mattermost/...server](https://github.com/mattermost/mattermost-server) [Rewrite existing plugin API tests to exclusively test via actual plugin](https://github.com/mattermost/mattermost-server/issues/10438)

#### Issue Link
https://github.com/mattermost/mattermost-plugin-github/issues/178

### CODE REVIEW

File: server/utils.go [hunk 1]
```diff
@@ -310,3 +308,27 @@ func getCodeMarkdown(user, repo, repoPath, word, lines string, isTruncated bool)
...
 }
+
+// getToDoDisplayText returns the text to be displayed in todo listings.
+func getToDoDisplayText(title, url, notifType string) string {
+	baseURL := "https://github.com/"
+
+	owner, repo := parseOwnerAndRepo(url, baseURL)
+	repoURL := fmt.Sprintf("%s%s/%s", baseURL, owner, repo)
+	repoWords := strings.Split(repo, "-")
+	if len(repoWords) != 1 {
```
Comment (levb) [inline server/utils.go]: maybe `if len(repo) > 20 && len(repoWords) > 1` - no reason to shorten names that are reasonably short?

File: server/utils_test.go [hunk 2]
```diff
@@ -67,12 +67,12 @@ func TestParseOwnerAndRepo(t *testing.T) {
 		{Full: "https://github.com/mattermost/mattermost-server", BaseURL: "", ExpectedOwner: "mattermost", ExpectedRepo: "mattermost-server"},
 		{Full: "https://github.com/mattermost/mattermost-server", BaseURL: "https://github.com/", ExpectedOwner: "mattermost", ExpectedRepo: "mattermost-server"},
 		{Full: "", BaseURL: "", ExpectedOwner: "", ExpectedRepo: ""},
-		{Full: "mattermost/mattermost/invalid_repo_url", BaseURL: "", ExpectedOwner: "", ExpectedRepo: ""},
-		{Full: "https://github.com/mattermost/mattermost/invalid_repo_url", BaseURL: "", ExpectedOwner: "", ExpectedRepo: ""},
+		{Full: "mattermost/mattermost/invalid_repo_url", BaseURL: "", ExpectedOwner: "mattermost", ExpectedRepo: "mattermost"},
```
Comment (levb) [inline server/utils_test.go]: 0/5 would be awesome to add tests for ... shortening

Comment (aayush-plivo) [reply server/utils_test.go]: Are you suggesting something like this:
```
{Full: "mattermost/...server", BaseURL: "", ExpectedOwner: "mattermost", ExpectedRepo: "...server"},
```

Comment (checkaayush) [reply server/utils_test.go]: Are you suggesting something like this:
```
{Full: "mattermost/...server", BaseURL: "", ExpectedOwner: "mattermost", ExpectedRepo: "...server"},
```

**Comment by jfrerich** [comment]
Thanks for the submission, @checkaayush!  I deployed your branch locally and will paste my `todo` items for reference. I see a couple of formatting issues.  Otherwise, awesome work!

* 

image

**Comment by jfrerich** [comment]
@aaronrothschild, @levb - what do y'all think about condensing the `Header Message` and the `submessages`?  They are kind of redundant. 

Examples
#### Unread Messages
##### You have 2 unread Messages:

#### Review Requests
##### You ahve 10 pull requests awaiting your review:

**Comment by checkaayush** [comment]
> Looks great, @checkaayush! I left some comments regarding formatting changes.

Made the suggested changes. 

Also, +1 for the idea of condensing header message and sub-messages. It looks way cleaner that way.

**Comment by jfrerich** [comment]
Thanks, @checkaayush!  For reference, I've updated to your commit and here is the result on my GH account
image

**Comment by checkaayush** [comment]
> @checkaayush, there are several items I think need feedback from @levb and @aaronrothschild. I like your implementation, but from a UI/UX, I think we are still a bit lacking. Perhaps this is for another ticket too. You don't need to implement these right now, but I would like to get feedback while in this ticket.
> 
> 1. I feel like the GH organization piece of the link is quite redundant.  Perhaps we could sort the tickets and add a level of bullet indentation for the `organization`, `repo`, `ticket title`
> 2. Perhaps add a `-` between `repo` and `ticket title`?  It's hard to differentiate.  Bolding is another option.
> 
> This is very subjective, but maybe we can get a quorum in this ticket. What do you think, @checkaayush?

Hmm, great input. How about we get rid of the sub-message, combine entries for the same repo together and sort them in descending order of number of items? No need to abridge repo names with ellipsis. For example, the _Review Requests_ section might look like this:

---------------------
### Review Requests

- [mattermost/mattermost-plugin-github](https://github.com/mattermost/mattermost-plugin-github)
  - [Result for /github subscribe list now shows flags](https://github.com/mattermost/mattermost-plugin-github/pull/197)
  - [Make links in todo messages consistent](https://github.com/mattermost/mattermost-plugin-github/pull/192)

- [mattermost/mattermost-server](https://github.com/mattermost/mattermost-server)
  - [Metrics operations shouldn't be inside goroutine](https://github.com/mattermost/mattermost-server/pull/13950)

**Comment by aaronrothschild** [comment]
One minor change:
 
 ### Review Requests (11)
 * [mattermost/mattermost-plugin-github](https://github.com/mattermost/mattermost-plugin-github)
   
   * [Result for /github subscribe list now shows flags](https://github.com/mattermost/mattermost-plugin-github/pull/197)
   * [Make links in todo messages consistent](https://github.com/mattermost/mattermost-plugin-github/pull/192)
 * [mattermost/mattermost-server](https://github.com/mattermost/mattermost-server)
   
   * [Metrics operations shouldn't be inside goroutine](https://github.com/mattermost/mattermost-server/pull/13950)

(note: I only added the "11" for the count of review requests pending)

**Comment by checkaayush** [comment]
@levb How do you feel about the suggested changes? Should I go ahead with them?

**Comment by jfrerich** [comment]
@checkaayush, I was on vacation since last Thursday, but back today.  I Like your proposed changes for several reasons:
* there is now a link to the repo, not just the individual ticket links
* the full title is shown
* tickets grouped by repo

If you also implement the changes suggest by @aaronrothschild, I believe this format is a winner!  Thank you so much for the new implementation.

File: server/utils_test.go (hunk 2 above)
Comment (levb) [reply server/utils_test.go]: Yes, exactly

**Comment by checkaayush** [comment]
> A personal preference would be to keep the PR # in the title, maybe "#231 the PR headline"?

@levb Just curious, any particular reason for this?

**Comment by levb** [comment]
> @levb Just curious, any particular reason for this?

not really, just used to see it, like in 
image

**Comment by checkaayush** [comment]
@jfrerich I took a small vacation and am yet to implement the suggested UX improvements. Although, it seems the changes till now have already been approved and tested. Were the suggested changes not in the scope of this issue? cc: @levb

**Comment by jfrerich** [comment]
Welcome back from your vacation, @checkaayush!  I hope you had a great time.  @levb has already approved, so I think you have successfully completed the PR request. 🎉  

Pinging @aaronrothschild for final review

**Comment by hanzei** [comment]
@aaronrothschild Any objections on merge this PR as it is?

**Pull request merged**